README.md
# Why the fork?

Hello there! You might like to consider [something apart from Nuts](https://github.com/electron/electron/blob/main/docs/tutorial/updates.md#deploying-an-update-server).

As [maintainer burnout is a thing](https://www.jeffgeerling.com/blog/2020/saying-no-burnout-open-source-maintainer), and the [original Nuts](https://github.com/GitbookIO/nuts) appears [to be abandoned](https://github.com/GitbookIO/nuts/issues/170) (thanks for all the great work!), this is a fork focused on upgrades as needed for general maintenance, reliability, security and on keeping the ability to download automatic updates by channel working when started locally (or deployed on a cloud provider, though I have not tested and have no intention to test Heroku) via the command line:

    # OWNER is a GitHub account, like pzrq
    # PRIVATE_REPO is any private repository with releases that need to be served publicly
    # TOKEN is most likely to be a GitHub server-to-server token, though other tokens might also work:
    #     https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/
    GITHUB_ENDPOINT=https://api.github.com GITHUB_REPO=$OWNER/$PRIVATE_REPO GITHUB_TOKEN=$TOKEN PORT=1031 npm start 

Then visiting in a browser to retrieve JSON and following any links to get a working direct download link:

    http://localhost:1031/update/channel/stable/windows/1.0.0

Other bits simply aren't mission critical to my use case so may be working, or may have broken - PRs are welcome to fix any regressions, though no new features will be added.

More details can be found in the [CHANGELOG.md](./CHANGELOG.md)

---

# Nuts

Nuts is a simple (and smart) application to serve desktop-application releases.

![Schema](./docs/schema.png)

It uses GitHub as a backend to store assets, and it can easily be deployed to Heroku as a stateless service. It supports GitHub private repositories (useful to store releases of a closed-source application available on GitHub).

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)

#### Features

- :sparkles: Store assets on GitHub releases
- :sparkles: Proxy releases from private repositories to your users
- :sparkles: Simple but powerful download urls
    - `/download/latest`
    - `/download/latest/:os`
    - `/download/:version`
    - `/download/:version/:os`
    - `/download/channel/:channel`
    - `/download/channel/:channel/:os`
- :sparkles: Support pre-release channels (`beta`, `alpha`, ...)
- :sparkles: Auto-updates with [Squirrel](https://github.com/Squirrel)
    - For Mac using `/update?version=<x.x.x>&platform=osx`
    - For Windows using Squirrel.Windows and Nugets packages
- :sparkles: Private API
- :sparkles: Use it as a middleware: add custom analytics, authentication
- :sparkles: Serve the perfect type of assets: `.zip` for Squirrel.Mac, `.nupkg` for Squirrel.Windows, `.dmg` for Mac users, ...
- :sparkles: Release notes endpoint
    - `/notes/:version`
- :sparkles: Up-to-date releases (GitHub webhooks)
- :sparkles: Atom/RSS feeds for versions/channels

#### Deploy it / Start it

[Follow our guide to deploy Nuts](https://nuts.gitbook.com/deploy.html).


#### Auto-updater / Squirrel

This server provides an endpoint for [Squirrel auto-updater](https://github.com/atom/electron/blob/master/docs/api/auto-updater.md), it supports both [OS X](https://nuts.gitbook.com/update-osx.html) and [Windows](https://nuts.gitbook.com/update-windows.html).

#### Documentation

[Check out the documentation](https://nuts.gitbook.com) for more details.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v4.0.0-pre.4](https://github.com/pzrq/nuts/compare/v4.0.0-pre.3...v4.0.0-pre.4)

- chore(deps): bump node-fetch from 2.6.6 to 2.6.7 [`#31`](https://github.com/pzrq/nuts/pull/31)
- chore(deps-dev): bump mocha from 9.1.4 to 9.2.0 [`#30`](https://github.com/pzrq/nuts/pull/30)
- chore(deps-dev): bump mocha from 9.1.3 to 9.1.4 [`#29`](https://github.com/pzrq/nuts/pull/29)

#### [v4.0.0-pre.3](https://github.com/pzrq/nuts/compare/v4.0.0-pre.2...v4.0.0-pre.3)

> 17 January 2022

- chore(deps): bump follow-redirects from 1.14.6 to 1.14.7 [`#28`](https://github.com/pzrq/nuts/pull/28)
- docs: Add electron's suggested alternatives [`1b26375`](https://github.com/pzrq/nuts/commit/1b263756519930425656403f19bde99d965c170b)

#### [v4.0.0-pre.2](https://github.com/pzrq/nuts/compare/v4.0.0-pre.1...v4.0.0-pre.2)

> 7 January 2022

- docs: Add explanation of fork and my goals [`f2a69d8`](https://github.com/pzrq/nuts/commit/f2a69d86f2b3c7685cfdcedd98aa0e0bf969146a)
- fix: TypeError: urljoin is not a function [`4b318fa`](https://github.com/pzrq/nuts/commit/4b318fa01c721e79f3848fa0dd16f9d71d8e18f8)
- fix: TypeError: Class constructor LRUCache cannot be invoked without 'new' [`8ac378b`](https://github.com/pzrq/nuts/commit/8ac378bc8dcae08e19d59609e945829c570dbf8a)

#### [v4.0.0-pre.1](https://github.com/pzrq/nuts/compare/v4.0.0-pre.0...v4.0.0-pre.1)

> 7 January 2022

- Bump tmp from 0.0.28 to 0.2.1 [`#24`](https://github.com/pzrq/nuts/pull/24)
- Bump strip-bom from 2.0.0 to 4.0.0 [`#27`](https://github.com/pzrq/nuts/pull/27)
- Bump basic-auth from 1.1.0 to 2.0.1 [`#25`](https://github.com/pzrq/nuts/pull/25)
- Bump feed from 0.3.1 to 4.2.2 [`#26`](https://github.com/pzrq/nuts/pull/26)
- Bump urljoin.js from 0.1.0 to 1.0.0 [`#22`](https://github.com/pzrq/nuts/pull/22)
- Bump lru-cache from 4.1.5 to 6.0.0 [`#23`](https://github.com/pzrq/nuts/pull/23)
- Bump fs-extra from 0.26.7 to 10.0.0 [`#21`](https://github.com/pzrq/nuts/pull/21)
- chore(dev-deps): Add auto-changelog [`7dcfe0d`](https://github.com/pzrq/nuts/commit/7dcfe0d6d96d2dcd0710199632c5c7488a721c36)
- chore: Ignore strip-bom v5 upgrade [`adc80ff`](https://github.com/pzrq/nuts/commit/adc80ff038d403e93ebc60ba58ec71ad412d92e5)

#### [v4.0.0-pre.0](https://github.com/pzrq/nuts/compare/3.1.1...v4.0.0-pre.0)

> 23 December 2021

- chore: Remove unused backends/github.js imports [`#18`](https://github.com/pzrq/nuts/pull/18)
- chore(dev-deps): Upgrade from mocha@^5 to ^9 [`#16`](https://github.com/pzrq/nuts/pull/16)
- chore(deps): Make everything a caret range [`#15`](https://github.com/pzrq/nuts/pull/15)
- BREAKING: Bump octocat from 0.10.2 to 1.2.1 - GITHUB_ENDPOINT env var now required [`#6`](https://github.com/pzrq/nuts/pull/6)
- chore(deps): Bump analytics-node from 1.2.2 to 6.0.0 [`#8`](https://github.com/pzrq/nuts/pull/8)
- chore(deps): npm audit fix [`#14`](https://github.com/pzrq/nuts/pull/14)
- chore(deps): Bump express-useragent from 0.1.9 to 1.0.15 [`#10`](https://github.com/pzrq/nuts/pull/10)
- chore(deps): Bump uuid from 2.0.1 to 8.3.2 [`#3`](https://github.com/pzrq/nuts/pull/3)
- chore(deps): Bump github-webhook-handler from 0.5.0 to 1.0.0 [`#5`](https://github.com/pzrq/nuts/pull/5)
- chore(deps): Bump semver from 5.0.1 to 7.3.5 [`#1`](https://github.com/pzrq/nuts/pull/1)
- chore(dev-deps): Bump should from 7.0.4 to 13.2.3 [`#2`](https://github.com/pzrq/nuts/pull/2)
- chore(deps): Bump lodash from 3.7.0 to 4.17.21 [`#9`](https://github.com/pzrq/nuts/pull/9)
- chore(deps): Bump q from 1.2.0 to 1.5.1 [`#4`](https://github.com/pzrq/nuts/pull/4)
- chore(deps): Bump request from 2.60.0 to 2.88.2 [`#7`](https://github.com/pzrq/nuts/pull/7)
- ci: Drop node 4 to 12 [`#13`](https://github.com/pzrq/nuts/pull/13)
- ci: Add node 16 [`#12`](https://github.com/pzrq/nuts/pull/12)
- ci: Drop node 0.10 and 0.11 [`#11`](https://github.com/pzrq/nuts/pull/11)
- remove unneeded heroku Procfile [`#128`](https://github.com/pzrq/nuts/pull/128)
- Add support for non-"stable" update release channels [`#95`](https://github.com/pzrq/nuts/pull/95)
- [#94] Drop 'latest' from URLs [`#99`](https://github.com/pzrq/nuts/pull/99)
- Updated Depreciation In Example [`#82`](https://github.com/pzrq/nuts/pull/82)
- Added travis ci support for modern node versions [`#88`](https://github.com/pzrq/nuts/pull/88)
- updated linux support [`#84`](https://github.com/pzrq/nuts/pull/84)
- added an ignore list for cloud foundry [`#85`](https://github.com/pzrq/nuts/pull/85)
- support VCAP_APP_PORT and GITHUB_ENDPOINT [`#1`](https://github.com/pzrq/nuts/pull/1)
- [#94] Drop 'latest' from URLs [`#94`](https://github.com/GitbookIO/nuts/issues/94)
- feat: Add yarn.lock [`7bfa9aa`](https://github.com/pzrq/nuts/commit/7bfa9aa1a7eb278fb8e6f3497eb9741b2500ca9b)
- chore(dev-deps): Add auto-changelog [`acfe420`](https://github.com/pzrq/nuts/commit/acfe420973da6915eeee490c2165946b94ef92b0)
- chore(deps): Force requests@^2.88.2 [`ad5da07`](https://github.com/pzrq/nuts/commit/ad5da0711bbbf2577b38547ebbc36b2744945f42)

#### [3.1.1](https://github.com/pzrq/nuts/compare/3.1.0...3.1.1)

> 19 May 2016

- fixes issue with a single release [`#72`](https://github.com/pzrq/nuts/pull/72)
- Add missing context reference [`#65`](https://github.com/pzrq/nuts/pull/65)
- Bump version to 3.1.1 [`261b736`](https://github.com/pzrq/nuts/commit/261b736164f0773c7576540d0d171acf2557d80a)

#### [3.1.0](https://github.com/pzrq/nuts/compare/3.0.0...3.1.0)

> 29 April 2016

- Expose manual refresh endpoint [`#58`](https://github.com/pzrq/nuts/pull/58)
- remove versions.list() from github backend [`#63`](https://github.com/pzrq/nuts/pull/63)
- Update github.md [`#53`](https://github.com/pzrq/nuts/pull/53)
- Allow TRUST_PROXY to pass JSON arguments to Express [`#60`](https://github.com/pzrq/nuts/pull/60)

lib/backends/backend.js
var _ = require('lodash');
var Q = require('q');
var path = require('path');
var os = require('os');
var destroy = require('destroy');
var LRU = require('../utils/lru-diskcache');
var streamRes = require('stream-res');
var Buffer = require('buffer').Buffer;

function Backend(nuts, opts) {
    this.cacheId = 0;
    this.nuts = nuts;
    this.opts = _.defaults(opts || {}, {
        // Folder to cache assets
        cache: path.resolve(os.tmpdir(), 'nuts'),

        // Cache configuration
        cacheMax: 500 * 1024 * 1024,
        cacheMaxAge: 60 * 60 * 1000,
    });

    // Create cache
    this.cache = LRU(opts.cache, {
        max: opts.cacheMax,
        maxAge: opts.cacheMaxAge
    });

    _.bindAll(this);
}

// Memoize a function
Backend.prototype.memoize = function(fn) {
    var that = this;

    return _.memoize(fn, function() {
        return that.cacheId+Math.ceil(Date.now()/that.opts.cacheMaxAge)
    });
};

// New release? clear cache
Backend.prototype.onRelease = function() {
    this.cacheId++;
};

// Initialize the backend
Backend.prototype.init = function() {
    this.cache.init();
    return Q();
};

// List all releases for this repository
Backend.prototype.releases = function() {

};

// Return stream for an asset
Backend.prototype.serveAsset = function(asset, req, res) {
    var that = this;
    var cacheKey = asset.id;

    function outputStream(stream) {
        var d = Q.defer();
        streamRes(res, stream, d.makeNodeResolver());
        return d.promise;
    }

    res.header('Content-Length', asset.size);
    res.attachment(asset.filename);

    // Key exists
    if (that.cache.has(cacheKey)) {
        return that.cache.getStream(cacheKey)
            .then(outputStream);
    }

    return that.getAssetStream(asset)
    .then(function(stream) {
        return Q.all([
            // Cache the stream
            that.cache.set(cacheKey, stream),

            // Send the stream to the user
            outputStream(stream)
        ]);
    });
};

// Return stream for an asset
Backend.prototype.getAssetStream = function(asset) {

};

// Return stream for an asset
Backend.prototype.readAsset = function(asset) {
    return this.getAssetStream(asset)
    .then(function(res) {
        var d = Q.defer();
        var output = Buffer([]);

        function cleanup() {
            destroy(res);
            res.removeAllListeners();
        }

        res.on('data', function(buf) {
                output = Buffer.concat([output, buf]);
            })
            .on('error', function(err) {
                cleanup();
                d.reject(err);
            })
            .on('end', function() {
                cleanup();
                d.resolve(output);
            });

        return d.promise

    })
};

module.exports = Backend;


lib/utils/lru-diskcache.js
var Q = require('q');
var _ = require('lodash');
var fs = require('fs-extra');
var path = require('path');
var crc = require('crc');
var lru = require('lru-cache');
var fsWriteStream = require('fswrite-stream');

function DiskCache(rootPath, options) {
    if(!(this instanceof DiskCache)) {
        return new DiskCache(rootPath, options);
    }

    // Root folder to cache files in
    this.path = rootPath;

    // Options
    this.options = _.defaults(options || {}, {
        // Maximum size of the cache
        max: 10*1024*1024,

        // If maxEntries is defined, max is not taken in consideration
        maxEntries: null,

        // Maximum age in ms
        maxAge: undefined,

        onError: function(err) {
            console.error(err.stack);
        }
    });

    // Bind methods
    _.bindAll(this, [
        'init',
        'reset',
        'has',
        'wait',
        'get',
        'set',
        'del',
        'size',
        'prune',
        '_filename',
        '_read',
        '_readStream',
        '_write',
        '_writeStream',
        '_writeData',
        '_dispose',
        '_length'
    ]);

    // Pending writes
    this.pending = {};

    // Setup LRU
    this.lru = new lru({
        max: this.options.maxEntries || this.options.max,
        maxAge: this.options.maxAge,
        dispose: this._dispose,
        length: this._length
    });
}

// Initialize and prepare the cache
DiskCache.prototype.init = function() {
    this.reset();
};

// Reset the whole cache
DiskCache.prototype.reset = function() {
    var that = this;

    if (fs.existsSync(this.path)) {
        fs.removeSync(this.path);
    }

    fs.mkdirsSync(this.path);
    this.lru.reset();
};

// Return true if key exists
DiskCache.prototype.has = function(key) {
    return this.lru.has(key);
};

// Wait's until a potential set is finished
// So you know when it's safe to write
DiskCache.prototype.wait = function(key) {
    // Return pending promise
    if(key in this.pending) {
        return this.pending[key];
    }

    // Update recentness
    this.lru.get(key);

    // No pending set, you're good to go
    return Q(key);
};

// Get data as a buffer or string
DiskCache.prototype.get = function(key, opts) {
    var that = this;
    opts = opts || {};

    return this.wait(key)
        .then(function(key) {
            return that._read(key, { encoding: opts.encoding });
        });
};

// Get a stream
DiskCache.prototype.getStream = function(key) {
    return this.wait(key)
        .then(this._readStream);
};

// Writes a data or stream to disk
DiskCache.prototype.set = function(key, dataOrStream) {
    var that = this;

    // Return pending promise promise
    if(key in this.pending) {
        return this.pending[key];
    }

    // Promise of writes etc ...
    var p = this._write(key, dataOrStream)
        .then(function(size) {
            // Mark key as saved
            that.lru.set(key, size);

            // Cleanup pending
            delete that.pending[key];

            // Return key
            return key;
        }, function(err) {
            // Remove key due to a failed write
            that.lru.del(key);

            // Cleanup pending
            delete that.pending[key];

            // Throw error
            throw err;
        });

    // Set promise as pending for now
    this.pending[key] = p;

    // Return promise
    return p;
};

// Deletes a key out of the cache.
DiskCache.prototype.del = function(key) {
    this.lru.del(key);
};

// Return total size of objects in cache taking into account
DiskCache.prototype.size = function() {
    return this.lru.length;
};

// Manually iterates over the entire cache proactively pruning old entries
DiskCache.prototype.prune = function() {
    return this.lru.prune();
};

// What's the filename of a key on the disk ?
DiskCache.prototype._filename = function(key) {
    return path.join(
        this.path,
        crc.crc32(key).toString(16)
    );
};

DiskCache.prototype._read = function(key, opts) {
    var d = Q.defer();

    fs.readFile(this._filename(key), opts || {}, d.makeNodeResolver());

    return d.promise;
};

DiskCache.prototype._readStream = function(key) {
    return Q(fs.createReadStream(this._filename(key)));
};


DiskCache.prototype._write = function(key, dataOrStream) {
    return (
        // Is this a stream ?
        (dataOrStream.pipe !== undefined) ?

            // If so, write as stream to disk
            this._writeStream(key, dataOrStream) :

            // Else, write as buffer to disk
            this._writeData(key, dataOrStream)
    );
};

DiskCache.prototype._writeStream = function(key, stream) {
    var d = Q.defer();

    fsWriteStream(this._filename(key), stream, d.makeNodeResolver());

    return d.promise;
};

DiskCache.prototype._writeData = function(key, data) {
    var d = Q.defer();
    var filename = this._filename(key);

    // Write data to disk
    fs.writeFile(filename, data, d.makeNodeResolver());

    return d.promise

        // Return size
        .then(function() {
            return Q.nfcall(fs.stat, filename)
                .get('size');
        });
};

//// Methods for lru-cache

DiskCache.prototype._dispose = function(key, value) {
    // If the file didn't go to disk there's nothing to destroy
    if(key in this.pending) {
        return;
    }

    try {
        fs.unlinkSync(this._filename(key));
    } catch(err) {
        this.options.onError(err);
    }
};

DiskCache.prototype._length = function(value, key) {
    // the value in memory is the size of the file
    return this.options.maxEntries? 1 : value;
};

module.exports = DiskCache;


lib/nuts.js
var _ = require('lodash');
var Q = require('q');
var Feed = require('feed');
var urljoin = require('urljoin.js').default;
var Understudy = require('understudy');
var express = require('express');
var useragent = require('express-useragent');

var BACKENDS = require('./backends');
var Versions = require('./versions');
var notes = require('./utils/notes');
var platforms = require('./utils/platforms');
var winReleases = require('./utils/win-releases');
var API_METHODS = require('./api');

function getFullUrl(req) {
    return req.protocol + '://' + req.get('host') + req.originalUrl;
}

function Nuts(opts) {
    if (!(this instanceof Nuts)) return new Nuts(opts);
    var that = this;

    Understudy.call(this);
    _.bindAll(this, [
    	'_init',
    	'performQ',
    	'serveAsset',
    	'onDownload',
    	'onUpdateRedirect',
    	'onUpdate',
    	'onUpdateWin',
    	'onServeNotes',
    	'onServeVersionsFeed',
    	'onAPIAccessControl']);

    this.opts = _.defaults(opts || {}, {
        // Backend to use
        backend: 'github',

        // Timeout for releases cache (seconds)
        timeout: 60*60*1000,

        // Pre-fetch list of releases at startup
        preFetch: true,

        // Secret for GitHub webhook
        refreshSecret: 'secret'
    });

    // .init() is now a memoized version of ._init()
    this.init = _.memoize(this._init);

    // Create router
    this.router = express.Router();

    // Create backend
    this.backend = new (BACKENDS(this.opts.backend))(this, this.opts);
    this.versions = new Versions(this.backend);

    // Bind routes
    this.router.use(useragent.express());

    this.router.get('/', this.onDownload);
    this.router.get('/download/channel/:channel/:platform?', this.onDownload);
    this.router.get('/download/version/:tag/:platform?', this.onDownload);
    this.router.get('/download/:tag/:filename', this.onDownload);
    this.router.get('/download/:platform?', this.onDownload);

    this.router.get('/feed/channel/:channel.atom', this.onServeVersionsFeed);

    this.router.get('/update', this.onUpdateRedirect);
    this.router.get('/update/:platform/:version', this.onUpdate);
    this.router.get('/update/channel/:channel/:platform/:version', this.onUpdate);
    this.router.get('/update/:platform/:version/RELEASES', this.onUpdateWin);
    this.router.get('/update/channel/:channel/:platform/:version/RELEASES', this.onUpdateWin);

    this.router.get('/notes/:version?', this.onServeNotes);

    // Bind API
    this.router.use('/api', this.onAPIAccessControl);
    _.each(API_METHODS, function(method, route) {
        that.router.get('/api/' + route, function(req, res, next) {
            return Q()
            .then(function() {
                return method.call(that, req);
            })
            .then(function(result) {
                res.send(result);
            }, next);
        });
    });
}

// _init does the real init work, initializing backend and prefetching versions
Nuts.prototype._init = function() {
    var that = this;
    return Q()
    .then(function() {
        return that.backend.init();
    })
    .then(function() {
        if (!that.opts.preFetch) return
        return that.versions.list();
    });
}


// Perform a hook using promised functions
Nuts.prototype.performQ = function(name, arg, fn) {
    var that = this;
    fn = fn || function() { };

    return Q.nfcall(this.perform, name, arg, function (next) {
        Q()
        .then(function() {
            return fn.call(that, arg);
        })
        .then(function() {
            next();
        }, next);
    })
};

// Serve an asset to the response
Nuts.prototype.serveAsset = function(req, res, version, asset) {
    var that = this;

    return that.init()
    .then(function() {
        return that.performQ('download', {
            req: req,
            version: version,
            platform: asset
        }, function() {
            return that.backend.serveAsset(asset, req, res)
        });
    });
};

// Handler for download routes
Nuts.prototype.onDownload = function(req, res, next) {
    var that = this;
    var channel = req.params.channel;
    var platform = req.params.platform;
    var tag = req.params.tag || 'latest';
    var filename = req.params.filename;
    var filetypeWanted = req.query.filetype;

    // When serving a specific file, platform is not required
    if (!filename) {
        // Detect platform from useragent
        if (!platform) {
            if (req.useragent.isMac) platform = platforms.OSX;
            if (req.useragent.isWindows) platform = platforms.WINDOWS;
            if (req.useragent.isLinux) platform = platforms.LINUX;
            if (req.useragent.isLinux64) platform = platforms.LINUX_64;
        }

        if (!platform) return next(new Error('No platform specified and impossible to detect one'));
    } else {
        platform = null;
    }

    // If specific version, don't enforce a channel
    if (tag != 'latest') channel = '*';

    this.versions.resolve({
        channel: channel,
        platform: platform,
        tag: tag
    })

    // Fallback to any channels if no version found on stable one
    .fail(function(err) {
        if (channel || tag != 'latest') throw err;

        return that.versions.resolve({
            channel: '*',
            platform: platform,
            tag: tag
        });
    })

    // Serve downloads
    .then(function(version) {
        var asset;

        if (filename) {
            asset = _.find(version.platforms, {
                filename: filename
            });
        } else {
            asset = platforms.resolve(version, platform, {
                wanted: filetypeWanted? '.'+filetypeWanted : null
            });
        }

        if (!asset) throw new Error("No download available for platform "+platform+" for version "+version.tag+" ("+(channel || "beta")+")");

        // Call analytic middleware, then serve
        return that.serveAsset(req, res, version, asset);
    })
    .fail(next);
};


// Request to update
Nuts.prototype.onUpdateRedirect = function(req, res, next) {
    Q()
    .then(function() {
        if (!req.query.version) throw new Error('Requires "version" parameter');
        if (!req.query.platform) throw new Error('Requires "platform" parameter');

        return res.redirect('/update/'+req.query.platform+'/'+req.query.version);
    })
    .fail(next);
};

// Updater used by OSX (Squirrel.Mac) and others
Nuts.prototype.onUpdate = function(req, res, next) {
    var that = this;
    var fullUrl = getFullUrl(req);
    var platform = req.params.platform;
    var channel = req.params.channel || '*';
    var tag = req.params.version;
    var filetype = req.query.filetype ? req.query.filetype : "zip";

    Q()
    .then(function() {
        if (!tag) throw new Error('Requires "version" parameter');
        if (!platform) throw new Error('Requires "platform" parameter');

        platform = platforms.detect(platform);

        return that.versions.filter({
            tag: '>='+tag,
            platform: platform,
            channel: channel
        });
    })
    .then(function(versions) {
        var latest = _.first(versions);
        if (!latest || latest.tag == tag) return res.status(204).send('No updates');

        var notesSlice = versions.slice(0, -1);
        if (versions.length === 1) {
            notesSlice = [versions[0]];
        }
        var releaseNotes = notes.merge(notesSlice, { includeTag: false });
        console.error(latest.tag);
        var gitFilePath = (channel === '*' ? '/../../../' : '/../../../../../');
        res.status(200).send({
            "url": urljoin(fullUrl, gitFilePath, '/download/version/'+latest.tag+'/'+platform+'?filetype='+filetype),
            "name": latest.tag,
            "notes": releaseNotes,
            "pub_date": latest.published_at.toISOString()
        });
    })
    .fail(next);
};

// Update Windows (Squirrel.Windows)
// Auto-updates: Squirrel.Windows: serve RELEASES from latest version
// Currently, it will only serve a full.nupkg of the latest release with a normalized filename (for pre-release)
Nuts.prototype.onUpdateWin = function(req, res, next) {
    var that = this;

    var fullUrl = getFullUrl(req);
    var platform = 'win_32';
    var channel = req.params.channel || '*';
    var tag = req.params.version;

    that.init()
    .then(function() {
        platform = platforms.detect(platform);

        return that.versions.filter({
            tag: '>='+tag,
            platform: platform,
            channel: channel
        });
    })
    .then(function(versions) {
        // Update needed?
        var latest = _.first(versions);
        if (!latest) throw new Error("Version not found");

        // File exists
        var asset = _.find(latest.platforms, {
            filename: 'RELEASES'
        });
        if (!asset) throw new Error("File not found");

       return that.backend.readAsset(asset)
       .then(function(content) {
            var releases = winReleases.parse(content.toString('utf-8'));

            releases = _.chain(releases)

                // Change filename to use download proxy
                .map(function(entry) {
                    var gitFilePath = (channel === '*' ? '../../../../' : '../../../../../../');
                    entry.filename = urljoin(fullUrl, gitFilePath, '/download/'+entry.semver+'/'+entry.filename);

                    return entry;
                })

                .value();

            var output = winReleases.generate(releases);

            res.header('Content-Length', output.length);
            res.attachment("RELEASES");
            res.send(output);
       });
    })
    .fail(next);
};

// Serve releases notes
Nuts.prototype.onServeNotes = function(req, res, next) {
    var that = this;
    var tag = req.params.version;

    Q()
    .then(function() {
        return that.versions.filter({
            tag: tag? '>='+tag : '*',
            channel: '*'
        });
    })
    .then(function(versions) {
        var latest = _.first(versions);

        if (!latest) throw new Error('No versions matching');

        res.format({
            'text/plain': function(){
                res.send(notes.merge(versions));
            },
            'application/json': function(){
                res.send({
                    "notes": notes.merge(versions, { includeTag: false }),
                    "pub_date": latest.published_at.toISOString()
                });
            },
            'default': function() {
                res.send(releaseNotes);
            }
        });
    })
    .fail(next);
};

// Serve versions list as RSS
Nuts.prototype.onServeVersionsFeed = function(req, res, next) {
    var that = this;
    var channel = req.params.channel || 'all';
    var channelId = channel === 'all'? '*' : channel;
    var fullUrl = getFullUrl(req);

    var feed = new Feed({
        id: 'versions/channels/'+channel,
        title: 'Versions (' + channel + ')',
        link: fullUrl
    });

    Q()
    .then(function() {
        return that.versions.filter({
            channel: channelId
        });
    })
    .then(function(versions) {
        _.each(versions, function(version) {
            feed.addItem({
                title: version.tag,
                link:  urljoin(fullUrl, '/../../../', '/download/version/'+version.tag),
                description: version.notes,
                date: version.published_at,
                author: []
            });
        });

        res.set('Content-Type', 'application/atom+xml; charset=utf-8');
        res.send(feed.render('atom-1.0'));
    })
    .fail(next);
};

// Control access to the API
Nuts.prototype.onAPIAccessControl = function(req, res, next) {
    this.performQ('api', {
        req: req,
        res: res
    })
    .then(function() {
        next();
    }, next);
};


module.exports = Nuts;
